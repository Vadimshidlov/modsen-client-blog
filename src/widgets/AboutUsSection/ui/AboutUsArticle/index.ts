export { AboutUsArticle } from '@/widgets/AboutUsSection/ui/AboutUsArticle/AboutUsArticle';
