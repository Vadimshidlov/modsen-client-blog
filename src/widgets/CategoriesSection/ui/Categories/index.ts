export { Categories } from '@/widgets/CategoriesSection/ui/Categories/Categories';
