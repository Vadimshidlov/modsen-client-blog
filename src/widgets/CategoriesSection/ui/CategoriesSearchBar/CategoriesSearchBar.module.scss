.searchBarContainer {
  position: relative;
  margin-bottom: 58px;

  @media (max-width: $mobile) {
    margin-bottom: 15px;
  }

  .searchInput {
    padding: 15px 84px 15px 25px;
    font-weight: $bold;
    font-size: $fz14;
    line-height: 24px;
    border: 1px solid $inputBorder;
    border-radius: 6px;
    outline: none;
    color: $black;

    &::placeholder {
      font-weight: $bold;
      font-size: $fz14;
      line-height: 24px;
      color: $black;
    }
  }

  .searchBarForm {
    position: relative;

    @media (max-width: 1250px) {
      max-width: 300px;
    }

    .searchButtonWrapper {
      position: absolute;
      right: 0;
      top: 0;
    }
  }

  .filteredPosts {
    padding-top: 5px;
    max-width: 300px;
    position: absolute;
    display: flex;
    flex-direction: column;
    z-index: 1;
    background: $white;

    .filteredPost {
      padding: 10px;
      border: 1px solid $grey;
      cursor: pointer;
      transition: all .5s ease;

      &:hover {
        background: $hoverYellow;
      }

      .filteredPostTitle {
        @include headingSix;
      }
      .filteredPostTags {
        @include headingSeven;
      }
    }

    .filteredPostsNotFound {
      width: 100%;
      @include headingSix;
      padding: 10px;
      border: 1px solid $grey;
      transition: all .5s ease;
    }
  }
}
