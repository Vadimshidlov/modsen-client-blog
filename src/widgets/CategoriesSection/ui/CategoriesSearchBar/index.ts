export { CategoriesSearchBar } from '@/widgets/CategoriesSection/ui/CategoriesSearchBar/CategoriesSearchBar';
