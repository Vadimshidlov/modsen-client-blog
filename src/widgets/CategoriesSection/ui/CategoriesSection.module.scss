.categoriesSection {
  display: flex;
  margin-bottom: 128px;
  padding-top: 102px;
  // column-gap: 147px;
  column-gap: 40px;
  justify-content: space-between;

  @media (max-width: 1250px) {
    flex-direction: column-reverse;
  }

  @media (max-width: $tablet) {
    padding-top: 51px;
    margin-bottom: 64px;
  }

  @media (max-width: $mobile) {
    padding-top: 25px;
    margin-bottom: 32px;
  }

  .postsContainer {
    display: flex;
    flex-direction: column;
    row-gap: 32px;
    max-width: 844px;

    .postsNotFountTitle {
      @include headingThree;

      @media (max-width: $tablet) {
        @include headingThree;
      }

      @media (max-width: $mobile) {
        @include headingFive;
      }
    }
  }

  .searchBarWrapper {
    max-width: 300px;
    display: flex;
    flex-direction: column;

    @media (max-width: 1250px) {
      max-width: 100%;
    }
  }
}
