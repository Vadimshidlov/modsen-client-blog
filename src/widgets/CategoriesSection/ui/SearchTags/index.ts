export { SearchTags } from '@/widgets/CategoriesSection/ui/SearchTags/SearchTags';
