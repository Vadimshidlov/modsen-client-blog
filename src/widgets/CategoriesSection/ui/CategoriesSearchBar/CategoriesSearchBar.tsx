'use client';

import React, { ChangeEvent, useMemo, useState } from 'react';
import Link from 'next/link';
import { useTranslations } from 'next-intl';
import { useDebounce } from '@/shared/utils';
import { Button } from '@/shared/ui/Button';
import { POSTS } from '@/shared/constants/posts';
import styles from '@/widgets/CategoriesSection/ui/CategoriesSearchBar/CategoriesSearchBar.module.scss';

export function CategoriesSearchBar() {
  const t = useTranslations('CategoriesSearchBar');

  const [searchValue, setSearchValue] = useState('');
  const searchValueDebounced = useDebounce(searchValue, 500);

  const handleSearValueChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    setSearchValue(value);
  };

  const filteredPosts = useMemo(
    () =>
      POSTS.filter((postData) =>
        postData.tags.some((tagData) =>
          tagData.toLowerCase().includes(searchValueDebounced.toLowerCase()),
        ),
      ),
    [searchValueDebounced],
  );

  return (
    <div className={styles.searchBarContainer}>
      <form action="submit" className={styles.searchBarForm}>
        <input
          className={styles.searchInput}
          type="text"
          placeholder={t('inputPlaceholder')}
          value={searchValue}
          onChange={handleSearValueChange}
          data-testid="searchTag-input"
        />
        <div className={styles.searchButtonWrapper}>
          <Button className={styles.searchButton} isPrimary={false}>
            {t('buttonText')}
          </Button>
        </div>
      </form>
      <ul className={styles.filteredPosts}>
        {searchValue !== '' &&
          filteredPosts.length !== 0 &&
          filteredPosts.slice(0, 5).map(({ title, tags, id }) => (
            <li className={styles.filteredPost} key={id}>
              <Link href={`/posts/${id}`} data-testid={`searched-post-${id}`}>
                <h3 className={styles.filteredPostTitle}>{title}</h3>
                <span className={styles.filteredPostTags}>{tags.join(', ')}</span>
              </Link>
            </li>
          ))}
        {searchValue !== '' && filteredPosts.length === 0 && (
          <li className={styles.filteredPostsNotFound} data-testid="not-found-posts">
            Posts not found
          </li>
        )}
      </ul>
    </div>
  );
}
