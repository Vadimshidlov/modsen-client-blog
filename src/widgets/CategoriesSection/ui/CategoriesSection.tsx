'use client';

import { useState } from 'react';
import styles from '@/widgets/CategoriesSection/ui/CategoriesSection.module.scss';
import container from '@/shared/styles/container.module.scss';
import { CategoriesSearchBar } from '@/widgets/CategoriesSection/ui/CategoriesSearchBar';
import { POSTS } from '@/shared/constants/posts';
import { PostItem } from '@/widgets/PostItem';
import { Categories } from '@/widgets/CategoriesSection/ui/Categories';
import { SearchTags } from '@/widgets/CategoriesSection/ui/SearchTags';
import { useTranslations } from 'next-intl';
import { CategoriesSectionPropsType } from '@/shared/types';

export function CategoriesSection({ selectedCategory }: CategoriesSectionPropsType) {
  const t = useTranslations('CategoriesSection');

  const [selectTags, setSelectTags] = useState<string[]>([]);

  const filteredPosts = POSTS.filter(
    (postData) =>
      postData.category.toLowerCase() === selectedCategory.toLowerCase() &&
      (selectTags.length === 0 || postData.tags.some((tag) => selectTags.includes(tag))),
  );

  return (
    <section className={`${styles.categoriesSection} ${container.container}`}>
      <div className={styles.postsContainer}>
        {filteredPosts.length === 0 && (
          <span className={styles.postsNotFountTitle}>{t('notFoundMessage')}</span>
        )}
        {filteredPosts.length !== 0 &&
          filteredPosts
            .slice(0, 3)
            .map(({ id, img, category, title, text }) => (
              <PostItem category={category} id={id} img={img} text={text} title={title} key={id} />
            ))}
      </div>
      <div className={styles.searchBarWrapper}>
        <SearchTags selectTags={selectTags} setSelectTags={setSelectTags} />
        <CategoriesSearchBar />
        <Categories />
      </div>
    </section>
  );
}
