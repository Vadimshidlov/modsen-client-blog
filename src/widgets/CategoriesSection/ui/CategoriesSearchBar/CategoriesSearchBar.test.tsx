import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { CategoriesSearchBar } from '@/widgets/CategoriesSection/ui/CategoriesSearchBar';

jest.mock('next-intl', () => ({
  useTranslations: jest.fn().mockReturnValue((key: string) => `Mocked ${key}`),
}));

describe('CategoriesSearchBar tests', () => {
  test('should matches snapshot', () => {
    const { container } = render(<CategoriesSearchBar />);
    expect(container).toMatchSnapshot();
  });

  test('should renders search input correctly', () => {
    const { getByTestId } = render(<CategoriesSearchBar />);

    expect(getByTestId('searchTag-input')).toBeInTheDocument();
  });

  test('should updates search value correctly', async () => {
    const { getByTestId } = render(<CategoriesSearchBar />);
    const searchInput = getByTestId('searchTag-input');

    fireEvent.change(searchInput, { target: { value: 'Tag1' } });
    await waitFor(() => expect(searchInput).toHaveValue('Tag1'));
  });

  test('should renders "Posts not found" message correctly', async () => {
    const { getByTestId } = render(<CategoriesSearchBar />);
    const searchInput = getByTestId('searchTag-input');

    fireEvent.change(searchInput, { target: { value: 'Tag4' } });

    await waitFor(() => {
      expect(getByTestId('not-found-posts')).toBeInTheDocument();
    });
  });
});
