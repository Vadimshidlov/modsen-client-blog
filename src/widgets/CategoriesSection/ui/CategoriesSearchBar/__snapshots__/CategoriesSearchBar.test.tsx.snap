// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesSearchBar tests should matches snapshot 1`] = `
<div>
  <div
    class="searchBarContainer"
  >
    <form
      action="submit"
      class="searchBarForm"
    >
      <input
        class="searchInput"
        data-testid="searchTag-input"
        placeholder="Mocked inputPlaceholder"
        type="text"
        value=""
      />
      <div
        class="searchButtonWrapper"
      >
        <button
          class="button borderedButton"
          type="button"
        >
          Mocked buttonText
        </button>
      </div>
    </form>
    <ul
      class="filteredPosts"
    />
  </div>
</div>
`;
